### DISCUSSION TITLE
Autostart

### DISCUSSION THREAD

**Discussion by tomlut** [tphakala/birdnet-go#136] [General]
What's the best way to have BirdNet-Go start on boot up, but in a terminal window rather than as a background service?

**Comment by tphakala** [comment]
Depends on a OS I guess, but why you need to start in terminal window instead of background service?

**Comment by tomlut** [comment]
Raspberry Pi OS. In a terminal window to see what is going on with a quick glance via VNC e.g. dog -> owl false positive rejection.

**Comment by tphakala** [comment]
Okay, in this case I would still start birdnet-go as a service, you can follow console output with journalctl.

Create a systemd config file /etc/systemd/system/birdnet-go.service, below is a config template you can use

```
[Unit]
Description=BirdNET-Go Analyzer Service

[Service]
ExecStart=/home/thakala/birdnet-go realtime
Restart=on-failure
User=thakala
Group=thakala
WorkingDirectory=/home/thakala

[Install]
WantedBy=multi-user.target
```

Replace **User, Group, ExecStart and WorkingDirectory** to match your system.

To reload systemd config run

`# systemctl daemon-reload`

Enable birdnet-go service start on boot run (systemctl start birdnet-go will start it on demand)

`# systemctl enable birdnet-go`

And to follow console output with journalctl run

`# journalctl -t birdnet-go -f`

**-t** matches logs from birdnet-go process, **-f** keeps following output and prints all log messages as they are received from birdnet-go

**Comment by ericleslie** [comment]
Not sure if this is best practice, but I set my docker container to always run which will automatically restart it on a crash or a system reboot. Added this to the runtime command.
`--restart=always`

**Reply by Vertikar** [reply]
Another option is --restart unless-stopped.

Always is fine too, but if you restart the Docker daemon the container will restart if you've manually stopped it.
details here - https://docs.docker.com/config/containers/start-containers-automatically/

**Comment by tomlut** [comment]
I can't get this to work. 
```
tom@raspberrypi:~ $ sudo systemctl status birdnet-go.service
x birdnet-go.service - BirdNET-Go Analyzer Service
     Loaded: loaded (/etc/systemd/system/birdnet-go.service; enabled; preset: e>
     Active: failed (Result: exit-code) since Sun 2024-05-05 10:22:16 AEST; 1h >
   Duration: 7ms
    Process: 703 ExecStart=/home/tom/birdnet/birdnet-go realtime (code=exited, >
   Main PID: 703 (code=exited, status=200/CHDIR)
        CPU: 3ms

May 05 10:22:16 raspberrypi (rdnet-go)[703]: birdnet-go.service: Failed at step>
May 05 10:22:16 raspberrypi systemd[1]: birdnet-go.service: Main process exited>
May 05 10:22:16 raspberrypi systemd[1]: birdnet-go.service: Failed with result >
May 05 10:22:16 raspberrypi systemd[1]: birdnet-go.service: Scheduled restart j>
May 05 10:22:16 raspberrypi systemd[1]: Stopped birdnet-go.service - BirdNET-Go>
May 05 10:22:16 raspberrypi systemd[1]: birdnet-go.service: Start request repea>
May 05 10:22:16 raspberrypi systemd[1]: birdnet-go.service: Failed with result >
May 05 10:22:16 raspberrypi systemd[1]: Failed to start birdnet-go.service - Bi>
```

The unzipped files are loacted here: `/home/tom/birdnet`


File /etc/systemd/system/birdnet-go.service contains:
```
[Unit]
Description=BirdNET-Go Analyzer Service

[Service]
ExecStart=/home/tom/birdnet/birdnet-go realtime
Restart=on-failure
User=tom
Group=tom
WorkingDirectory=/home/tom/birdnet

[Install]
WantedBy=multi-user.target
```

Works perfectly fine if I run `./birdnet-go realtime` from the install folder.

**Reply by tphakala** [reply]
Although your provided log lines are truncated and are missing some key information, (code=exited, status=200/CHDIR) indicates that your working directory setting is incorrect. Please check your install directory path.

**Comment by tomlut** [comment]
That's the full log. And `/home/tom/birdnet` is the correct directory.

**Comment by tomlut** [comment]
Is there a permission problem for the soundcard?

```
tom@raspberrypi:~ $ journalctl -t birdnet-go -f
May 06 14:14:29 raspberrypi birdnet-go[1194]:     8: PulseAudio Sound Server, 70756c7365,
May 06 14:14:29 raspberrypi birdnet-go[1194]:     9: Plugin using Speex DSP (resample, agc, denoise, echo, dereverb), 7370656578,
May 06 14:14:29 raspberrypi birdnet-go[1194]:     10: Plugin for channel upmix (4,6,8), 75706d6978,
May 06 14:14:29 raspberrypi birdnet-go[1194]:     11: Plugin for channel downmix (stereo) with a simple spacialization, 76646f776e6d6978,
May 06 14:14:29 raspberrypi birdnet-go[1194]:     12: USB Device 0x41e:0x30d3, USB Audio, 3a332c30,
May 06 14:14:29 raspberrypi birdnet-go[1194]: ALSA lib pcm_asym.c:105:(_snd_pcm_asym_open) capture slave is not defined
May 06 14:14:29 raspberrypi birdnet-go[1194]: ALSA lib pcm_dsnoop.c:566:(snd_pcm_dsnoop_open) unable to open slave
May 06 14:14:29 raspberrypi birdnet-go[1194]: ALSA lib pcm_dsnoop.c:566:(snd_pcm_dsnoop_open) unable to open slave
May 06 14:14:29 raspberrypi birdnet-go[1194]: ALSA lib pcm_dsnoop.c:566:(snd_pcm_dsnoop_open) unable to open slave
May 06 14:14:29 raspberrypi birdnet-go[1194]: 2024/05/06 14:14:29 Device init failed miniaudio: Failed to open backend device
^C
tom@raspberrypi:~ $ sudo systemctl status birdnet-go.service
x birdnet-go.service - BirdNET-Go Analyzer Service
     Loaded: loaded (/etc/systemd/system/birdnet-go.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Mon 2024-05-06 14:14:29 AEST; 1min 8s ago
   Duration: 308ms
    Process: 1194 ExecStart=/home/tom/birdnet/birdnet-go realtime (code=exited, status=1/FAILURE)
   Main PID: 1194 (code=exited, status=1/FAILURE)
        CPU: 313ms

May 06 14:14:29 raspberrypi systemd[1]: birdnet-go.service: Scheduled restart job, restart counter is at 5.
May 06 14:14:29 raspberrypi systemd[1]: Stopped birdnet-go.service - BirdNET-Go Analyzer Service.
May 06 14:14:29 raspberrypi systemd[1]: birdnet-go.service: Start request repeated too quickly.
May 06 14:14:29 raspberrypi systemd[1]: birdnet-go.service: Failed with result 'exit-code'.
May 06 14:14:29 raspberrypi systemd[1]: Failed to start birdnet-go.service - BirdNET-Go Analyzer Service.
tom@raspberrypi:~ $ 

```

When started manually it works:
```
BirdNET GLOBAL 6K V2.4 FP32 model initialized, using 4 threads of available 4 CPUs
System details: linux raspbian 12.5 on Raspberry Pi 4 Model B Rev 1.5 hardware
Starting analyzer in realtime mode. Threshold: 0.8, sensitivity: 1, interval: 15
2024/05/06 14:16:47 Connecting to MQTT broker
2024/05/06 14:16:47 Connected to MQTT broker: tcp://10.1.1.100:1883
2024/05/06 14:16:47 Successfully connected to MQTT broker
[2024-05-06T14:16:47+10:00] [INFO] Ã¢Â¨ http server started on [::]:8080
Capture Devices
    0: Discard all samples (playback) or generate zero samples (capture), 6e756c6c, 
    1: Playback/recording through the PulseAudio sound server, 64656661756c74, [default]
    2: Default Audio Device, 73797364656661756c74, 
    3: Rate Converter Plugin Using Libav/FFmpeg Library, 6c617672617465, 
    4: Rate Converter Plugin Using Samplerate Library, 73616d706c6572617465, 
    5: Rate Converter Plugin Using Speex Resampler, 737065657872617465, 
    6: JACK Audio Connection Kit, 6a61636b, 
    7: Open Sound System, 6f7373, 
    8: PulseAudio Sound Server, 70756c7365, 
    9: Plugin using Speex DSP (resample, agc, denoise, echo, dereverb), 7370656578, 
    10: Plugin for channel upmix (4,6,8), 75706d6978, 
    11: Plugin for channel downmix (stereo) with a simple spacialization, 76646f776e6d6978, 
    12: USB Device 0x41e:0x30d3, USB Audio, 3a332c30, 
Listening on device: Playback/recording through the PulseAudio sound server (device ID 64656661756c74)
```

**Reply by farski** [reply]
It's been a while, so I don't know if I was having the exact same problem, but I also was having a bunch of weird permissions issues when trying to set up a systemd service for this. Once the work on Docker deploys started, I didn't spend any more time on it because once things stabilize I will likely use that. For now I've just been running `birdnet-go realtime &`. I check the web UI often enough that I notice pretty quickly if it's gone down. But I tried a whole bunch of different things (sudo, not sudo, keeping it in `/opt` vs `/home`, different users and groups, etc) and never got it working the way I wanted. But I also don't know my way around systemd very well, so I may have been missing something very obvious.

**Reply by tphakala** [reply]
@tomlut thank you, now there is enough logs so I was able to debug, I installed full RPI OS and was able to reproduce this. Problem is caused by Raspberry PI's integrated sound device and fact that you have full Raspberry Pi install, I have not hit this as I am using RPi OS Lite install and by habit I always disable unused devices, which is kind of poor practice when you try to develop something for others to use ..



There are few fixes, if you do not need Raspberry Pi's audio playback just disable onboard audio devices. Edit /boot/firmware/config.txt 

change dtparam=audio=on to off

`dtparam=audio=off`

add noaudio to dtoverlay=vc4-kms-v3d

`dtoverlay=vc4-kms-v3d,noaudio`

and reboot.

If you need to use onboard audio playback add create .asoundrc file in your home director (/home/tom) and add following content

```
pcm.!default {
  type asym
  playback.pcm {
    type plug
    slave.pcm "hw:0,0"
  }
  capture.pcm {
    type plug
    slave.pcm "hw:2,0"
  }
}

ctl.!default {
  type hw
  card 2
}
```

This will make second audio card (which should be your USB Sound Card) as default capture device.

I will try to fix birdnet-go code so that it would always prefer USB Sound Card's over RPI sound device.

**Comment by tomlut** [comment]
I do not need audio so I gave the first option a go. 

Worked perfectly. Thank you. 👍
